src/main/java/frc/robot/subsystems/Vision.java

```diff
@@ -84,18 +87,90 @@ public PhotonTrackedTarget getBestTarget() {
      */
     public boolean isLinedUp() {
 
-        // Get the best target
-        PhotonTrackedTarget bestTarget = getBestTarget();
+        // Check to see if the yaw deviation is below the tolerance
+        return (Parameters.vision.YAW_TOLERANCE > getBestTarget().getYaw());
+    }
 
-        // Make sure that the target isn't null (that we have a target)
-        if (bestTarget != null) {
+    // return list of list for parseTargetCorners
+    private List<List<TargetCorner>> getTargetCorners() {
+        PhotonPipelineResult pipelineResult = camera.getLatestResult();
+        if (!pipelineResult.hasTargets()) return null;
 
-            // Check to see if the yaw deviation is below the tolerance
-            return (Parameters.vision.YAW_TOLERANCE > Math.abs(bestTarget.getYaw()));
-        } else {
-            // We don't have a target, so we're not lined up
-            return false;
+        List<List<TargetCorner>> ret = new ArrayList<>();
+        List<PhotonTrackedTarget> targets = pipelineResult.getTargets();
+
+        for (PhotonTrackedTarget target : targets) {
+            ret.add(target.getCorners());
         }
+
+        return ret;
+    }
+
+    private List<TargetCorner> parsedTargetCorners() {
+        List<List<TargetCorner>> cornerData = getTargetCorners();
+        if (cornerData == null) return null;
+
+        List<TargetCorner> ret = new ArrayList<>();
+
+        // assumes target contours are rectangles
+        for (List<TargetCorner> corners : cornerData) {
+            TargetCorner p1 = corners.get(0);
+            TargetCorner p2 = corners.get(1);
+            TargetCorner p3 = corners.get(2);
+            TargetCorner p4 = corners.get(3);
+
+            // following canvas axises (down is larger)
+            double midY = (p1.y + p2.y + p3.y + p4.y) / 4;
+
+            if (p1.y < midY) ret.add(p1);
```

Comment:
Here, you should account for upper and lower corners as they have different heights. Maybe put them into their own list or you can account for it later.

Reply:
I'm not sure what you mean. Here I'm just taking the upper corners and removing the lower corners, because they'll be mapped to the same global position. I can take the average of their heights if that's what you mean

Reply:
I get what you mean now, that makes sense.